docs/Changelog.md

```diff
@@ -29032,6 +29032,97 @@ This version of the operator has been available since version 23 of the default
+  For cross attention, query and key might have different lengths.
+
+  This operator also covers the 3 following variants based on the number of heads:
+  1) Multi-headed Attention (MHA): Described in the paper https://arxiv.org/pdf/1706.03762, q_num_heads = kv_num_heads.
+  2) Group-query Attention (GQA): Described in the paper https://arxiv.org/pdf/2305.13245, q_num_heads > kv_num_heads.
+  3) Multi-query Attention (MQA): Described in the paper https://arxiv.org/pdf/1911.02150, q_num_heads > kv_num_heads, q_num_heads=1.
+
+  Attention bias to be added is calculated based on attn_mask input and is_causal attribute, only one of which can be provided.
+  1) If is_causal is set to 1, the attention masking is a lower triangular matrix when the mask is a square matrix. The attention masking has the form of the upper left causal bias due to the alignment.
+  2) attn_mask: A boolean mask where a value of True indicates that the element should take part in attention or a float mask of the same type as query, key, value that is added to the attention score.
+
+  Both past and present state key/values are optional. They shall be used together, and not allowed to use only one of them.
+  The following pattern is applied to the Q, K and V inputs after appropriate reshaping of K and V inputs based on sequence lengths and num heads provided:
+
+            Q          K          V
+            |          |          |
+            |      Transpose      |
+            |          |          |
+            ---MatMul---          |
+                  |               |
+         scale---Mul              |
+                  |               |
+       at_bias---Add              |
+                  |               |
+               Softmax            |
+                  |               |
+                  -----MatMul------
+                          |
+                          Y
+
+
+#### Version
+
+This version of the operator has been available since version 23 of the default ONNX operator set.
+
+#### Attributes
+
+<dl>
+<dt><tt>is_causal</tt> : int (default is 0)</dt>
+<dd>If set to 1, the attention masking is a lower triangular matrix when the mask is a square matrix. The attention masking has the form of the upper left causal bias due to the alignment.</dd>
+<dt><tt>kv_num_heads</tt> : int</dt>
+<dd>Number of heads of key and value. Must use with for 3D inputs of Q, K and V. </dd>
+<dt><tt>q_num_heads</tt> : int</dt>
+<dd>Number of heads of query. Must use with for 3D inputs of Q, K and V. </dd>
+<dt><tt>scale</tt> : float</dt>
+<dd>Scaling factor applied prior to softmax. Default value is 1/sqrt(head_size)</dd>
+</dl>
+
+#### Inputs (3 - 6)
+
+<dl>
+<dt><tt>Q</tt> : T</dt>
+<dd>Query tensor. 4D tensor with shape (batch_size, q_num_heads, q_sequence_length, head_size) or 3D tensor with shape (batch_size, q_sequence_length, q_hidden_size). For cases with a 3D input tensor, q_hidden_size = q_num_heads * head_size</dd>
```

Comment:
I am trying to figure out why specify 4D or 3D tensors as inputs when the 3D tensor contains the same information as the 4D tensor with some of it hidden.

I we only supported 4D tensors we would not need the kv_num_heads and q_num_heads attributes because they would be baked into the shape of the QKV tensors.

If we only used 4D for all inputs the question becomes what is the best shape?
(batch_size, q_num_heads, q_sequence_length, head_size) or (batch_size, q_sequence_length, q_num_heads, head_size) or some other.

I have not read all the GQA or MQA papers yet so don't know if there is a reason for both shapes based on the papers.

Not suggesting a change just trying to understand the need for both 3d and 4d tensor shapes if they both contain the same info.

Reply:
Good question. Worth thinking through and simplifying this. I wonder if 3D is useful for the special-case where num-heads is 1 (though it is only a minor matter to reshape to add a 1-valued dimension for the num-heads).